src/new/wallet.js

```diff
@@ -0,0 +1,441 @@
+    const allTokens = 'allTokens' in walletData ? walletData.allTokens : [];
+
+    let isNewTx = true;
+    if (newTx.tx_id in historyTransactions) {
+      isNewTx = false;
+    }
+
+    wallet.updateHistoryData(historyTransactions, allTokens, [newTx], null, walletData);
+
+    if (isNewTx) {
+      this.emit('new-tx', newTx);
+    } else {
+      this.emit('update-tx', newTx);
+    }
+    return;
+
+    // TODO XXX Uncomment the following block to keep track of the balance.
+    /*
+    // We need to reload it because it was modified by updateHistoryData.
+    const newWalletData = wallet.getWalletData();
+    const { keys } = newWalletData;
+
+    const updatedHistoryMap = {};
+    const updatedBalanceMap = {};
+    const balances = this.getTxBalance(newTx);
+
+    // we now loop through all tokens present in the new tx to get the new history and balance
+    for (const [tokenUid, tokenTxBalance] of Object.entries(balances)) {
+      // we may not have this token yet, so state.tokensHistory[tokenUid] would return undefined
+      const currentHistory = state.tokensHistory[tokenUid] || [];
+      const newTokenHistory = addTxToSortedList(tokenUid, tx, tokenTxBalance, currentHistory);
+      updatedHistoryMap[tokenUid] = newTokenHistory;
+      // totalBalance should not be confused with tokenTxBalance. The latter is the balance of the new
+      // tx, while the former is the total balance of the token, considering all tx history
+      const totalBalance = getBalance(tokenUid);
+      updatedBalanceMap[tokenUid] = totalBalance;
+    }
+    const newTokensHistory = Object.assign({}, state.tokensHistory, updatedHistoryMap);
+    const newTokensBalance = Object.assign({}, state.tokensBalance, updatedBalanceMap);
+
+    storage.setItem('local:tokens:history', newTokensHistory);
+    storage.setItem('local:tokens:balance', newTokensBalance);
+    */
+  };
+
+  /**
+   * Send a transaction with multi tokens
+   * @param {Object} data Array of {'address', 'value', 'token'}
+   */
+  sendMultiTokenTransaction(data) {
+    const txData = {
+      inputs: [],
+      outputs: []
+    }
+
+    const walletData = wallet.getWalletData();
+    const historyTxs = 'historyTransactions' in walletData ? walletData.historyTransactions : {};
+
+    // First I need an array with all tokens
+    const allTokens = [];
+    for (const d of data) {
+      if (allTokens.indexOf(d.token) === -1) {
+        allTokens.push(d.token);
+      }
+    }
+
+    for (const d of data) {
+      const dataOutput = {'address': d.address, 'value': d.value, 'tokenData': tokens.getTokenIndex(allTokens, d.token.uid)};
+      const partialData = {'inputs': [], 'outputs': [dataOutput]}
+      const result = wallet.prepareSendTokensData(partialData, d.token, true, historyTxs, allTokens);
+
+      if (!result.success) {
+        console.log('Error sending tx:', result.message);
+        return;
```

Comment:
Why not throw an exception?

Reply:
I am returning a `Promise.reject` now, since this method returns promises. What do you think?